pkg/image/admission/imagepolicy/accept.go

```diff
@@ -0,0 +1,111 @@
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+var errRejectByPolicy = fmt.Errorf("this image is prohibited by policy")
+
+func accept(accepter rules.Accepter, resolver imageResolver, m meta.ImageReferenceMutator, attr admission.Attributes, covered policyAttributes, excludedRules sets.String) error {
+	gr := attr.GetResource().GroupResource()
+	requiresImage := accepter.RequiresImage(gr)
+	resolvesImage := accepter.ResolvesImage(gr)
+
+	errs := m.Mutate(func(image string) (string, error) {
+		ok, newImage, err := acceptImage(accepter, resolver, image, gr, covered, requiresImage, resolvesImage, excludedRules)
+		if err != nil {
+			return image, err
+		}
+		if !ok {
+			return image, errRejectByPolicy
+		}
+		return newImage, nil
+	})
+
+	for i := range errs {
+		if errs[i].Detail == errRejectByPolicy.Error() {
+			errs[i].Type = field.ErrorTypeForbidden
+		} else {
+			errs[i].Detail = fmt.Sprintf("this image is prohibited by policy: %s", errs[i].Detail)
+		}
+	}
+
+	if len(errs) > 0 {
+		glog.V(5).Infof("failed to create: %v", errs)
+		return apierrs.NewInvalid(attr.GetKind().GroupKind(), attr.GetName(), errs)
+	}
+	glog.V(5).Infof("allowed: %#v", attr)
+	return nil
+}
+
+func acceptImage(accepter rules.Accepter, resolver imageResolver, name string, gr unversioned.GroupResource, covered policyAttributes, requiresImage, resolvesImage bool, excludedRules sets.String) (bool, string, error) {
+	// only resolve the image once per pod
+	if covered.Has(name) {
+		return true, name, nil
+	}
+	// explicitly allow empty strings to bypass checks
+	if len(name) == 0 {
+		return true, name, nil
+	}
+
+	ref, err := imageapi.ParseDockerImageReference(name)
+	if err != nil {
+		return false, name, err
+	}
+
+	// check the decision cache if we've already made a decision on the image
+	// TODO: time bound the decision cache so that namespace overrides can take effect
+	isDigest := len(ref.ID) > 0
+
+	var image *imageapi.Image
+	originalName := name
+	if requiresImage || (!isDigest && resolvesImage) {
+		image, err = resolver.Resolve(ref)
+		switch {
+		case apierrs.IsNotFound(err) && !resolvesImage:
+			// if the referenced image does not exist, and resolution is not required, we'll let
+			// the policy rule decide whether to fail
+		case err != nil:
+			return false, name, err
+		default:
+			glog.V(5).Infof("Resolved image %v to %s (%s)", ref, image.Name, image.DockerImageReference)
+			// resolve the reference to a digest, rather than a tag
+			ref.Tag = ""
+			ref.ID = image.Name
+
+			// if we should resolve the image, update the image and also check the cache again
+			if resolvesImage {
+				name = ref.Exact()
+				isDigest = true
+			}
+		}
+	}
+
+	attr := &rules.ImagePolicyAttributes{
+		Name:          ref,
+		OriginalName:  name,
+		Resource:      gr,
+		Image:         image,
+		ExcludedRules: excludedRules,
+	}
+	// remember under both names
+	covered.Remember(name, attr)
+	covered.Remember(originalName, attr)
```

Comment:
I think this causes problems when you're trying to force resolution.  You can end up looking up the old name and not getting the new name you normally return.